### DISCUSSION TITLE
implementing-custom-token-providers-for-passwordless-authentication-in-asp-net-core-identity/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#255] [Announcements]
# implementing-custom-token-providers-for-passwordless-authentication-in-asp-net-core-identity/

In this post I show how to create 2 custom token providers for ASP.NET Core Identity, so that you can implement passwordless login with a short token expiry

https://andrewlock.net/implementing-custom-token-providers-for-passwordless-authentication-in-asp-net-core-identity/

**Comment by davidhenley** [comment]
Great post! Solved exactly what I needed for a custom refresh token lifetime.

Thank you!

**Comment by LordBlackWolf** [comment]
Hey this is great but how do we get an access token and include user claims etc?

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at April 11 2018, 05:48, in reply to [Muhammad Rehan Saeed](https://disqus.com/by/RehanSaeedUK/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/255#discussioncomment-6045867)</em>

---

<p>In terms of security, I agree they're more secure than 1fa: you must already control an email account; can't use credential stuffing attacks. You can always combine them with 2fa to improve security too. They're especially good at protecting non-tech people I think, as you can't reuse the same crap password you do everywhere else!</p><p>You're not the only one though - I do find them annoying too, especially when compared to a well-integrated password manager.</p>

[a comment by an automated account is not shown]

**Comment by danieltlk** [comment]
Great article, thank you!

Can u please explain why "TOTP tokens are good for tokens with very short lifetimes (nominally 30 seconds)"?
Thanks!

**Reply by andrewlock** [reply]
Because TOTP tokens are only valid for 30 seconds +/- a margin of error 🙂

**Comment by AlexZeitler** [comment]
Thanks, exactly what I've been looking for
